internal/services/v1/permissions.go

```diff
@@ -535,86 +536,157 @@ func (ps *permissionServer) LookupSubjects(req *v1.LookupSubjectsRequest, resp v
-		foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
-		if !ok {
-			return fmt.Errorf("missing resource ID in returned LS")
+	var currentCursor *dispatch.Cursor
+	remainingConcreteLimit := 0
+
+	lsRequestHash, err := computeLSRequestHash(req)
+	if err != nil {
+		return ps.rewriteError(ctx, err)
+	}
+
+	if req.OptionalCursor != nil {
+		decodedCursor, err := cursor.DecodeToDispatchCursor(req.OptionalCursor, lsRequestHash)
+		if err != nil {
+			return ps.rewriteError(ctx, err)
 		}
+		currentCursor = decodedCursor
+	}
+
+	if req.OptionalConcreteLimit > 0 {
+		remainingConcreteLimit = int(req.OptionalConcreteLimit)
+	}
+
+	internalResponseCursor := &impl.DecodedCursor{
+		VersionOneof: &impl.DecodedCursor_V1{
+			V1: &impl.V1Cursor{
+				Revision:              atRevision.String(),
+				CallAndParametersHash: lsRequestHash,
+			},
+		},
+	}
+
+	for {
+		countSubjectsFound := 0
 
-		for _, foundSubject := range foundSubjects.FoundSubjects {
-			excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+		ctxWithCancel, cancel := context.WithCancel(ctx)
+		defer cancel()
+
+		stream := dispatchpkg.NewHandlingDispatchStream(ctxWithCancel, func(result *dispatch.DispatchLookupSubjectsResponse) error {
+			foundSubjects, ok := result.FoundSubjectsByResourceId[req.Resource.ObjectId]
+			if !ok {
+				return fmt.Errorf("missing resource ID in returned LS")
 			}
 
-			excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
-			for _, excludedSubject := range foundSubject.ExcludedSubjects {
-				resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+			for _, foundSubject := range foundSubjects.FoundSubjects {
+				// Skip wildcards if requested they be skipped.
+				if req.WildcardOption == v1.LookupSubjectsRequest_WILDCARD_OPTION_EXCLUDE_WILDCARDS && foundSubject.SubjectId == tuple.PublicWildcard {
+					continue
+				}
+
+				excludedSubjectIDs := make([]string, 0, len(foundSubject.ExcludedSubjects))
+				excludedSubjects := make([]*v1.ResolvedSubject, 0, len(foundSubject.ExcludedSubjects))
+				for _, excludedSubject := range foundSubject.ExcludedSubjects {
+					excludedSubjectIDs = append(excludedSubjectIDs, excludedSubject.SubjectId)
+
+					resolvedExcludedSubject, err := foundSubjectToResolvedSubject(ctx, excludedSubject, caveatContext, ds)
+					if err != nil {
+						return fmt.Errorf("error when resolving excluded subject: %w", err)
+					}
+
+					if resolvedExcludedSubject == nil {
+						continue
+					}
+
+					excludedSubjects = append(excludedSubjects, resolvedExcludedSubject)
+				}
+
+				subject, err := foundSubjectToResolvedSubject(ctx, foundSubject, caveatContext, ds)
```

Comment:
shouldn't we be computing this before we resolve caveats to the excluded subjects? If resolving the caveat of the found subject == `nil`, then there is no need to compute the excluded subjects since `nil - anything = nil`. You are in fact short-circuiting already right after that.

Reply:
Not sure what you mean here

Reply:
I'm suggesting we can move:

https://github.com/josephschorr/spicedb/blob/0c9fbb7c8f9570fbb38364b8db17612071d2cce4/internal/services/v1/permissions.go#L603-L609

Above

https://github.com/josephschorr/spicedb/blob/0c9fbb7c8f9570fbb38364b8db17612071d2cce4/internal/services/v1/permissions.go#L586-L601

That way we don't have to compute the caveats for the `ExcludedSubjects` if the first statement returns `nil`.

Reply:
oooooh. good catch